libagent/win_server.py

```diff
@@ -0,0 +1,357 @@
+        self.close()
+
+    def settimeout(self, timeout):
+        self.sock.settimeout(timeout)
+
+    def recv(self, size):
+        try:
+            with ctrl_cancel_async_io(self.sock.fileno()):
+                return self.sock.recv(size)
+        except OSError as e:
+            if e.winerror == 10054:
+                # Convert socket close to end of file
+                return None
+            raise
+
+    def sendall(self, reply):
+        with ctrl_cancel_async_io(self.sock.fileno()):
+            return self.sock.sendall(reply)
+
+    def close(self):
+        return self.sock.close()
+
+    def getsockname(self):
+        return self.sock.getsockname()
+
+
+class Server:
+    def __init__(self, pipe_name):
+        self.timeout = None
+        self.pipe_name = pipe_name
+        self.sock = None
+        self.pipe = None
+        if not isinstance(self.pipe_name, str):
+            # GPG simulated socket via localhost socket
+            self.key = os.urandom(16)
+            self.sock = socket.socket()
+            self.sock.bind(('127.0.0.1', 0))
+            _, port = self.sock.getsockname()
+            self.sock.listen(1)
+            # Write key to file
+            with open(self.pipe_name, 'wb') as f:
+                with ctrl_cancel_async_io(f.fileno()):
+                    f.write(str(port).encode())
+                    f.write(b'\n')
+                    f.write(self.key)
+
+    def __del__(self):
+        if self.pipe is not None:
+            self.pipe.close()
+        self.pipe = None
+        if self.sock is not None:
+            self.sock.close()
+        self.sock = None
+
+    def settimeout(self, timeout):
+        if self.sock:
+            self.sock.settimeout(timeout)
+        self.timeout = timeout
+
+    def getsockname(self):
+        return self.pipe_name
+
+    def accept(self):
+        if self.sock:
+            with ctrl_cancel_async_io(self.sock.fileno()):
+                sock, addr = self.sock.accept()
+            sock = InterruptibleSocket(sock)
+            sock.settimeout(self.timeout)
+            if self.key != util.recv(sock, 16):
+                sock.close()
+                # Simulate timeout on failed connection to allow the caller to retry
+                raise TimeoutError('Illegitimate client tried to connect to pipe {0}'
+                                   .format(self.name))
+            sock.settimeout(None)
+            return (sock, addr)
+        else:
+            # Named pipe based server
+            if self.pipe is None:
+                self.pipe = NamedPipe.create(self.pipe_name)
+            self.pipe.settimeout(self.timeout)
+            self.pipe.connect()
+            self.pipe.settimeout(None)
+            # A named pipe can only accept a single connection
+            # It must be recreated if a new connection is to be made
+            pipe = self.pipe
+            self.pipe = None
+            return (pipe, self.pipe_name)
+
+
+class Client:
```

Comment:
Please explain the logic behind using both a socket and a named pipe here.

Reply:
Tool dependent. Windows doesn't have AF_UNIX sockets (well, it does since Windows 10, but it is still rarely used), so different services ported from Linux circumvent this issue in different ways. OpenSSH uses named pipes. GPG uses a localhost socket, locked by a nonce that is written to a file. `win_server` automatically switches between the two based on whether the socket path is a string (coming from an argument like `--sock-path) or byte string (return value from `gpg.keyring.get_agent_sock_path`)

A more robust way may be to check for `\\.\pipe\` prefix. But given that the addresses already come in two different types, they would have to be normalized first.